core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -314,23 +363,17 @@ protected void responseWithFailure(Throwable t) {
 
         @Override
         protected void doRun() throws Exception {
-            try (Releasable shardReference = getIndexShardOperationsCounter(request.internalShardId)) {
-                shardOperationOnReplica(request.internalShardId, request);
+            assert request.shardId() != null : "request shardId must be set";
+            try (Releasable shardReference = getIndexShardOperationsCounter(request.shardId())) {
```

Comment:
fixed in https://github.com/areek/elasticsearch/commit/f2d6a931a9fdd82b85c543f9d68be1e1f70399a0